### PR TITLE
Unify input checks on detection models

### PR DESCRIPTION
**Pull request by mmcenta**
Follows up on #2207, unifies checks of the training targets in detection models, and adds checks for all fields (boxes and labels for all models, masks for MaskRCNN, and keypoints for KeypointRCNN).

I followed @fmassa's suggestion to move the validation inside the `roi_heads.py` file, which already contained a few checks. I merged the existing validation and added functionality.

Note that this can still be improved: I do not check that the masks have the right width and height or that the number of keypoints matches what the model expects.

### CODE REVIEW

File: torchvision/models/detection/roi_heads.py [hunk 1]
```diff
@@ -479,6 +479,50 @@ def paste_masks_in_image(masks, boxes, img_shape, padding=1):
...
+            the number of dimensions and any non-None values of this tuple.
+        shape_string (Optional[str]): optional string for exception messages.
+            If messages. If not specified, shape will be cast into string and
+            used instead.
+
+        Raises:
+            ValueError if the Tensor fails a check.
+        """
+    if not isinstance(dtype, tuple):
+        dtype = (dtype,)
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py:503]: torchscript doesn't support arguments with multiple types, so it might be better to always let `dtype` be a tuple here

File: torchvision/models/detection/roi_heads.py [hunk 2]
```diff
@@ -479,6 +479,50 @@ def paste_masks_in_image(masks, boxes, img_shape, padding=1):
...
+    if not shape_string:
+        shape_string = str(shape_string)
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py:505]: Was this meant to be `shape_string = str(shape)`?

File: torchvision/models/detection/roi_heads.py [hunk 3]
```diff
@@ -479,6 +479,50 @@ def paste_masks_in_image(masks, boxes, img_shape, padding=1):
     return ret
 
 
+def check_target_item(target, key, dtype, shape, shape_string=None):
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py:482]: Can you prefix this function with an underscore so that it is private? `_check_target_item`

Comment (mmcenta) [reply torchvision/models/detection/roi_heads.py:482]: Of course.

File: torchvision/models/detection/roi_heads.py (hunk 1 above)
Comment (fmassa) [reply torchvision/models/detection/roi_heads.py:503]: This is tricky.

I know that passing `int` works for dtype, but I think this might be a historical artifact.

Also, I would recommend using `List[int]` instead of a `Tuple`, as a tuple needs a fixed (known) size.


cc @eellison can we pass something else as `int` for the dtype type annotation in torchscript?

Comment (eellison) [reply torchvision/models/detection/roi_heads.py:503]: Fix in here https://github.com/pytorch/pytorch/pull/39741

**Comment by fmassa** [comment]
@mmcenta looking at it again, I was wrong :-)

The shape of the keypoints is indeed `[N, K, 3]` I believe. so if you could fix the tests, it would be great!

**Comment by mmcenta** [comment]
@fmassa Ok I will commit the changes.

Are you sure? What about the code in `coco_utils.py` with the different order?

**Comment by fmassa** [comment]
> Are you sure? What about the code in coco_utils.py with the different order?

I think I have misread the implementation before because of a bad variable name (`num_keypoints1), because `keypoints.shape[0]` is the number of annotations (objects).

**Comment by fmassa** [comment]
Also, can you rebase your changes on top of latest master? There seems to be conflicts.

**Comment by mmcenta** [comment]
@fmassa Sure, and sorry for all the commits, I need to figure out what CircleCI is complaining about before I can fix it. Before it was casting an Optional[str] to bool, now it is the `any` builtin (which I should have imagine wouldn't be allowed because there you used a method instead of all).

I will keep doing this until everything is fixed and then squash the commits and rebase to the new master :P

**Comment by mmcenta** [comment]
@fmassa I'm having trouble when comparing an Optional[int] to an int. I tried to assert like we did in the last PR but it didn't work. Do you have an idea of how we fix it?

**Comment by fmassa** [comment]
@mmcenta I think you need to do something like
```python
item = shape[i]
assert item is not None
```

**Comment by fmassa** [comment]
So there seems to be an issue with CI today https://github.com/pytorch/vision/pull/2316
Don't worry about failing tests, let me try to understand what is going on with CI and then we can continue checking what the issues are.

But there are some conflicts on your branch that could be fixed already.

**Comment by mmcenta** [comment]
@fmassa I think the checks that are failing now are not my fault (finally). Also sorry for the mess with the commit I think I messed up the rebase the first time around.

**Comment by fmassa** [comment]
Hi @mmcenta 

Sorry for the delay in getting back to you.

Can you rebase your PR on top of master?
Just doing `git pull origin master` (where `origin` points to `github.com/pytorch/vision`) should do it.

**Comment by mmcenta** [comment]
Done! There are some failing checks but I don't think it is a consequence of this PR.

**Comment by NicolasHug** [comment]
Hi @mmcenta , thanks for your work so far! Would you be interested in continuing this PR? There are a few linting issues, and merging with master should hopefully fix the unrelated CI issues, which would help for reviews. Thanks!
